SingularityService/src/main/java/com/hubspot/singularity/data/SingularityValidator.java

```diff
@@ -355,13 +364,117 @@ public SingularityDeploy checkDeploy(SingularityRequest request, SingularityDepl
 
     checkBadRequest(deployHistoryHelper.isDeployIdAvailable(request.getId(), deployId), "Can not deploy a deploy that has already been deployed");
 
+    if (deploy.getRunImmediately().isPresent()) {
+      deploy = checkImmediateRunDeploy(
+          request,
+          deploy,
+          deploy.getRunImmediately().get(),
+          activeTasks,
+          pendingTasks);
+    }
+
     if (request.isDeployable()) {
       checkRequestForPriorityFreeze(request);
     }
 
     return deploy;
   }
 
+  private SingularityDeploy checkImmediateRunDeploy(SingularityRequest request,
+                                                    SingularityDeploy deploy,
+                                                    SingularityRunNowRequest runNowRequest,
+                                                    List<SingularityTaskId> activeTasks,
+                                                    List<SingularityPendingTaskId> pendingTasks) {
+    if (!request.isScheduled() && !request.isOneOff()) {
+      throw badRequest("Can not request an immediate run of a non-scheduled / always running request (%s)", request);
+    }
+
+    boolean canRunImmediately =
```

Comment:
do we want to validate this at the time the deploy is enqueued? We re-check similar conditions at the time the pending request is created when the deploy finishes, and the data could have changed by then. Checking here also causes a few extra zk calls (activeTaskIds/pendingTaskIds) that could possibly be hit on the non-leader instance